kernel/src/main/java/org/kframework/backend/kore/ModuleToKORE.java

```diff
@@ -1113,6 +1140,12 @@ private void convert(Sort sort, Production prod) {
     }
 
     private void convert(Sort sort, boolean var) {
+        if (sort.name().equals(AddSortInjections.SORTPARAM_NAME)) {
```

Comment:
I have no idea why we have a "SortParam" sort but it seems very strange and hackish to me. What are you trying to accomplish?

Reply:
I hope to catch you on slack later and discuss this more interactively, but to give you a head-start, here is the issue I'm trying to address:


Well, the issue here is that ml opeations are parameterized.  Some of them (And, Or, Implies) have the sort of their operands, and perhaps for those it the poly mechanism might work, but others like Ceil and Equals, transport the truth from one sort to another, so in some sense they are injections.

the issue becomes even weirder when I have a combination of them, e.g. Ceil(Ceil(X:Int)) becomes \ceil{Q,R}(\ceil{Int, Q}(X)), where Q is a new sort parameter.  And of course, if any of Ceil, Equals, True or False are at the top, they would need a sort parameter as their output sort.

That is why I thought of annotating the ML constructs with their corresponding sort parameters during AddSortInhections.

Reply:
I should think Ceil(Ceil(X:Int)) would just become `\ceil{Int, R}(\ceil{Int,Int}(X))`. Then you wouldn't need new sort parameters. Is there a reason this is theoretically insufficient?

Reply:
tI think this would be unacceptable because then the pattern, being too specialized, might not match (and be able to be replaced by its RHS) in all contexts in which it would match.
Consider the rule motivating this PR:
`#Ceil(I1:Int /Int I2:Int) => #Equals(I2 =/=Int 0, true)`

Using the existing mechanism `#Ceil(I1:Int /Int I2:Int)` would become `\ceil{Int,Int}(I1:Int /Int I2:Int)` while `#Equals(I2 =/=Int 0, true)` would become `\equals{Bool,Bool}(I2 =/=Int 0, true)`.  However, I need to put an `\equals` on top of both of them, so if I leave the `\ceil` as Int and the `\equals` as bool, I would need to add inj to both of them.

In this particular case,the final result I currently generate (and which is the desirable one) is `\equals{Q, R}(\ceil{Int,Q}(I1:Int /Int I2:Int)), \equals{Bool, Q}(I2 =/=Int 0, true))`

But, as said before, you might get even more parameters, say if you want to express something like 
`#Equals(#Equals(#Not(#False), #True), #True)`, which should become something like `\equals{R,S}(\equals{Q,R}(\not{Q}(\bottom{Q}), \top{Q}), \top{R})`

Reply:
What if we go the other way? have it be `\equals{R, R}(\ceil{Int,R}(...), \equals(Bool,R}(...))` Does that have the same problem?